create_grid_with_stream.cu

```diff
@@ -0,0 +1,392 @@
+    // the location in the grid
+    Point *bl, *br, *tl, *tr;
+    bl = (Point *)malloc(h_grid_counts[0] * sizeof(Point));
+    br = (Point *)malloc(h_grid_counts[1] * sizeof(Point));
+    tl = (Point *)malloc(h_grid_counts[2] * sizeof(Point));
+    tr = (Point *)malloc(h_grid_counts[3] * sizeof(Point));
+
+    // Shift the data from device to host
+    cudaMemcpy(bl, bottom_left, h_grid_counts[0] * sizeof(Point),
+               cudaMemcpyDeviceToHost);
+    cudaMemcpy(br, bottom_right, h_grid_counts[1] * sizeof(Point),
+               cudaMemcpyDeviceToHost);
+    cudaMemcpy(tl, top_left, h_grid_counts[2] * sizeof(Point),
+               cudaMemcpyDeviceToHost);
+    cudaMemcpy(tr, top_right, h_grid_counts[3] * sizeof(Point),
+               cudaMemcpyDeviceToHost);
+
+    quad_q->push(Quad_point_info(bl, h_grid_counts[0], Quad(mp(x1, y1), mp(middle_x, middle_y))));
+    quad_q->push(Quad_point_info(br, h_grid_counts[1], Quad(mp(middle_x, y1), mp(x2, middle_y))));
+    quad_q->push(Quad_point_info(tl, h_grid_counts[2], Quad(mp(x1, middle_y), mp(middle_x, y2))));
+    quad_q->push(Quad_point_info(tr, h_grid_counts[3], Quad(mp(middle_x, middle_y), mp(x2, y2))));
+
+    // Free data
+    cudaFree(d_points);
+    cudaFree(d_categories);
+    cudaFree(d_grid_counts);
+    cudaFree(bottom_left);
+    cudaFree(bottom_right);
+    cudaFree(top_left);
+    cudaFree(top_right);
+
+    return;
+}
+
+void build_quadtree_levels(Point *points, int point_count, queue<Quad_point_info> *quad_q)
+{
+
+    // limit stream to 32 but even then not gauranteed that perormance will be good becaus eof the limited number of SMs
+    // limit by 4
+    double time_taken;
+    clock_t start, end;
+
+    pair<int, int> initial_bl = mp(0, 0);
+    pair<int, int> initial_tr = mp(1000, 1000);
+    Quad initial_grid = Quad(initial_bl, initial_tr);
+    start = clock();
+    quadtree_grid(points, point_count, initial_bl, initial_tr, nullptr, quad_q);
+    while (!quad_q->empty())
+    {
+        // start 4 streams at a time, one for each bl, br, tl, tr points
+        cudaStream_t streams[4] = {nullptr, nullptr, nullptr, nullptr};
+        for (int i = 0; i < 4; i++)
+        {
+            if (quad_q->empty()) break;
+
+            Quad_point_info quad_point_info = quad_q->front();
+            quad_q->pop();
+
+            int x1 = quad_point_info.region.bl.fi, y1 = quad_point_info.region.bl.se,
+                x2 = quad_point_info.region.tr.fi, y2 = quad_point_info.region.tr.se;
+            if (!(quad_point_info.count_in_quadrant < MIN_POINTS or (abs(x1 - x2) < MIN_DISTANCE and abs(y1 - y2) < MIN_DISTANCE)))
+            {
+                cudaStreamCreate(&(streams[i]));
+                printf("Stream %d created \n", i);
+                quadtree_grid(quad_point_info.quadrant_start_ptr, quad_point_info.count_in_quadrant, quad_point_info.region.bl, quad_point_info.region.tr, streams[i], quad_q);
+            }
+            else{
```

Comment:
This code is incomplete. Please change the condition

Reply:
Yep. Typo. Will not be there in next Revision